Added missing lines

mnabideltares opened the pull request:
**📝 Description**
Missing lines in Moh-Coulomb are added.

mnabideltares committed:
Added missing lines

mnabideltares committed:
Added missing lines

mnabideltares committed:
Added an integration test for colomn under gravity

mnabideltares committed:
Added an integration test for colomn under gravity

WPK4FEM commented on applications/GeoMechanicsApplication/tests/test_mohr_coulomb_with_tension_cutoff/test_column_under_gravity/ProjectParameters.json:
```diff
@@ -0,0 +1,151 @@
+        "newmark_gamma":                      0.5,
+        "newmark_theta":                      0.5,
+        "rayleigh_m":                         0.0,
+        "rayleigh_k":                         0.0,
```
Not used, as the computation has solution_type Quasi-Static

avdg81 replied:
Done.

WPK4FEM commented on applications/GeoMechanicsApplication/tests/test_mohr_coulomb_with_tension_cutoff/test_column_under_gravity/ProjectParameters.json:
```diff
@@ -0,0 +1,151 @@
+        "strategy_type":                      "newton_raphson",
+        "convergence_criterion":              "residual_criterion",
+        "displacement_relative_tolerance":    1.0E-9,
+        "displacement_absolute_tolerance":    1.0E-9,
```
Not used, as the convergence_criterion is residual_criterion.

avdg81 replied:
Done.

WPK4FEM commented on applications/GeoMechanicsApplication/tests/test_mohr_coulomb_with_tension_cutoff/test_column_under_gravity/ProjectParameters.json:
```diff
@@ -0,0 +1,151 @@
+        "second_alpha_value":                 1.0,
+        "min_alpha":                          0.1,
+        "max_alpha":                          2.0,
+        "line_search_tolerance":              0.5,
```
Not used, as the strategy_type is newton_raphson.

WPK4FEM commented on applications/GeoMechanicsApplication/tests/test_mohr_coulomb_with_tension_cutoff/test_column_under_gravity/ProjectParameters.json:
```diff
@@ -0,0 +1,151 @@
+        "min_alpha":                          0.1,
+        "max_alpha":                          2.0,
+        "line_search_tolerance":              0.5,
+        "rotation_dofs":                      true,
```
Why true, there are no elements with rotation D.o.F. in this computation.

WPK4FEM commented on applications/GeoMechanicsApplication/tests/test_mohr_coulomb_with_tension_cutoff/test_column_under_gravity/ProjectParameters.json:
```diff
@@ -0,0 +1,151 @@
+        "body_domain_sub_model_part_list":    ["Soil"]
+    },
+    "output_processes": {
+        "gid_output": [{
```
The test uses get_on_integration_points to retrieve effective stresses. None of the gid_output is used for comparision.

WPK4FEM reviewed (commented):
Please make the test files as short as possible.

mnabideltares committed:
Modifications based on review comments

WPK4FEM committed:
Merge branch 'geo/add-missing-lines' of https://github.com/KratosMultiphysics/Kratos into geo/add-missing-lines

WPK4FEM committed:
rotation_dofs false

WPK4FEM committed:
project parameters

WPK4FEM committed:
Test for constitutive matrix computation.

WPK4FEM committed:
Correction in README for source files and typo hexahedra.

WPK4FEM committed:
Unit tests for further coverage MC

WPK4FEM committed:
Merge remote-tracking branch 'origin/master' into geo/add-missing-lines

WPK4FEM committed:
Fast test suite i.s.o. nightly.

WPK4FEM committed:
correct unit test for RequiresInitializeMaterialResponse.

WPK4FEM committed:
corrected integration test result for doing 1 step only

WPK4FEM committed:
const to anonymous namespace.
Use KRATOS_EXPECT_TRUE i.s.o. KRATOS_EXPECT_EQ.

WPK4FEM committed:
Fixed GetLawFeatures implementation and added unit test.

WPK4FEM committed:
The KRATOS_EXPECT_TRUE macro should not be followed by a ;

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -209,12 +225,13 @@ void MohrCoulombWithTensionCutOff::InitializeMaterialResponseCauchy(Parameters&
-    rFeatures.SetOptions(mpConstitutiveDimension->GetSpatialType());
-    rFeatures.SetOptions(INFINITESIMAL_STRAINS);
-    rFeatures.SetOptions(ISOTROPIC);
+    Flags cl_options;
```
Perhaps simply `options` is good enough here? (I generally don't like abbreviations like `cl`.)

avdg81 replied:
Done.

WPK4FEM committed:
Removed empty table from MaterialParameters.json

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -209,12 +225,13 @@ void MohrCoulombWithTensionCutOff::InitializeMaterialResponseCauchy(Parameters&
+    Flags cl_options;
+    cl_options.Set(mpConstitutiveDimension->GetSpatialType());
+    cl_options.Set(ConstitutiveLaw::INFINITESIMAL_STRAINS, true);
+    cl_options.Set(ConstitutiveLaw::ISOTROPIC, true);
```
It appears we can leave out the value `true`, which is simpler and would make it more similar to the statement at line 229.

avdg81 replied:
Done.

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_constitutive/test_mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -19,6 +19,7 @@
 #include "tests/cpp_tests/test_utilities.h"
 
 #include <boost/numeric/ublas/assignment.hpp>
+#include <custom_constitutive/three_dimensional.h>
```
For non-system headers, we use double quotes rather than angular brackets:
```suggestion
#include "custom_constitutive/three_dimensional.h"
```

Also, I'd suggest to keep the "local" headers together. In other words, I'd move this line to just below line 19.

avdg81 replied:
Done.

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_constitutive/test_mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -388,4 +389,135 @@ KRATOS_TEST_CASE_IN_SUITE(MohrCoulombWithTensionCutOff_Serialization, KratosGeoM
+    properties.SetValue(GEO_TENSILE_STRENGTH, 10.0);
+    ConstitutiveLaw::Parameters parameters;
+    parameters.SetMaterialProperties(properties);
+    parameters.SetOptions(ConstitutiveLaw::COMPUTE_CONSTITUTIVE_TENSOR);
```
Perhaps it's clearer to use the mechanism that _adds_ options rather than _overwriting_ them, i.e.:
```suggestion
    parameters.GetOptions().Set(ConstitutiveLaw::COMPUTE_CONSTITUTIVE_TENSOR);
```

avdg81 replied:
Done.

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_constitutive/test_mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -388,4 +389,135 @@ KRATOS_TEST_CASE_IN_SUITE(MohrCoulombWithTensionCutOff_Serialization, KratosGeoM
+    // Act
+    Vector     cauchy_stress_vector = ZeroVector(4);
+    const auto dummy_process_info   = ProcessInfo{};
+    law.SetValue(CAUCHY_STRESS_VECTOR, cauchy_stress_vector, dummy_process_info);
```
I'm not sure whether we'd need this function call for this test. Do you know? (It continues to pass when I comment it out.)

avdg81 replied:
Done.

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_constitutive/test_mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -388,4 +389,135 @@ KRATOS_TEST_CASE_IN_SUITE(MohrCoulombWithTensionCutOff_Serialization, KratosGeoM
+    auto mc_law = MohrCoulombWithTensionCutOff(std::make_unique<PlaneStrain>());
+
+    // Act & Assert
+    KRATOS_EXPECT_EQ(mc_law.WorkingSpaceDimension(), 2);
```
Perhaps we'd also like to assert/show when this constitutive law is used in a 3D model, the returned value equals 3?
```suggestion
    KRATOS_EXPECT_EQ((MohrCoulombWithTensionCutOff{std::make_unique<PlaneStrain>()}.WorkingSpaceDimension()), 2);
    KRATOS_EXPECT_EQ((MohrCoulombWithTensionCutOff{std::make_unique<ThreeDimensional>()}.WorkingSpaceDimension()), 3);
```

We could consider to extend this idea to some of the other tests as well.

avdg81 replied:
Done.

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_constitutive/test_mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -388,4 +389,135 @@ KRATOS_TEST_CASE_IN_SUITE(MohrCoulombWithTensionCutOff_Serialization, KratosGeoM
+    KRATOS_EXPECT_TRUE(features.GetOptions().Is(ConstitutiveLaw::INFINITESIMAL_STRAINS))
+    KRATOS_EXPECT_TRUE(features.GetOptions().Is(ConstitutiveLaw::ISOTROPIC))
+    KRATOS_EXPECT_TRUE(features.GetOptions().Is(ConstitutiveLaw::THREE_DIMENSIONAL_LAW))
+    KRATOS_EXPECT_EQ(features.GetStrainMeasures().front(),ConstitutiveLaw::StrainMeasure_Infinitesimal);
```
I think we need to rerun `kp format`, because I reckon the formatting of this line does not comply with what Clang-format does.

avdg81 replied:
Done.

avdg81 reviewed (commented):
I'm very pleased to see the additional (unit) tests, which truly helps to build confidence in our code. I'm also see happy to see that the documentation has been improved and that the code has been cleaned up a bit. I have several minor suggestions, none of them being blocking.

avdg81 committed:
Processed review suggestions

- Added a few more assertions to cover 3D models.
- Renamed a local variable.
- Removed two redundant values.
- Use a better way to set constitutive law parameter options (avoid overwriting previously set options).
- Removed an unneeded function call.
- Applied consistent formatting to `#include`.
- Reformatted the C++ code using Clang-format.

avdg81 committed:
Processed other review suggestions

- Removed redundant Rayleigh parameters.
- Removed redundant displacement-based convergence criteria.

avdg81 committed:
Removed two empty input lists

rfaasse commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -209,12 +225,13 @@ void MohrCoulombWithTensionCutOff::InitializeMaterialResponseCauchy(Parameters&
-    rFeatures.SetOptions(INFINITESIMAL_STRAINS);
-    rFeatures.SetOptions(ISOTROPIC);
+    auto options = Flags{};
+    options.Set(mpConstitutiveDimension->GetSpatialType());
```
Just for my understanding: does this mean that before the options were overwritten?

avdg81 replied:
Yes, that is correct. @WPK4FEM discovered this while he was writing unit tests.

rfaasse commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_constitutive/test_mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -388,4 +389,121 @@ KRATOS_TEST_CASE_IN_SUITE(MohrCoulombWithTensionCutOff_Serialization, KratosGeoM
+
+KRATOS_TEST_CASE_IN_SUITE(MohrCoulombWithTensionCutOff_HasIncrementalFormulation, KratosGeoMechanicsFastSuiteWithoutKernel)
+{
+    EXPECT_TRUE(MohrCoulombWithTensionCutOff{std::make_unique<PlaneStrain>()}.IsIncremental());
+    EXPECT_TRUE(MohrCoulombWithTensionCutOff{std::make_unique<ThreeDimensional>()}.IsIncremental());
```
Since these are not really dependent on the ConstitutiveDimension, it doesn't add any coverage to check it twice. I'm fine with keeping it, but it might make more sense to only do one check (same holds for the rest of the tests where the results do not depend on the dimension)

avdg81 replied:
Indeed, it doesn't add anything to the code coverage, but it does communicate that some member functions return results that depend on the constitutive law dimension that was passed to the constructor, and other member functions return results that are independent of that. From a documentation point of view, I'd like to keep these additional assertions. Also note that the test case names reflect what to expect. What do you think?

rfaasse replied:
For me its fine either way, so let's keep it like this

rfaasse commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_constitutive/test_mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -388,4 +389,121 @@ KRATOS_TEST_CASE_IN_SUITE(MohrCoulombWithTensionCutOff_Serialization, KratosGeoM
+    KRATOS_EXPECT_EQ(features.GetStrainMeasures().front(), ConstitutiveLaw::StrainMeasure_Infinitesimal);
+    KRATOS_EXPECT_EQ(features.GetStrainSize(), 6);
+    KRATOS_EXPECT_EQ(features.GetSpaceDimension(), 3);
+}
```
Personally I'd split this into two tests (then it's more clear when e.g. one of them fails)

avdg81 replied:
Done.

rfaasse commented on applications/GeoMechanicsApplication/tests/test_mohr_coulomb_with_tension_cutoff/README.md:
```diff
@@ -23,4 +23,4 @@ The boundary conditions are shown below:
 
-These tests are also extended to three dimensional elements. The 3D cases are applied on hexahydrons of 8 nodes, namely 3D8N.
\ No newline at end of file
+These tests are also extended to three dimensional elements. The 3D cases are applied on 8 node hexahedra, namely 3D8N.
```
We could also add a few sentences about the new test (since it's a bit different than the rest)

avdg81 replied:
Done.

rfaasse commented on applications/GeoMechanicsApplication/tests/test_mohr_coulomb_with_tension_cutoff/test_column_under_gravity/ProjectParameters.json:
```diff
@@ -0,0 +1,139 @@
+                        "node_output":         false,
+                        "skin_output":         false,
+                        "nodal_results":       ["DISPLACEMENT","NODAL_CAUCHY_STRESS_TENSOR"],
+                        "gauss_point_results": ["GREEN_LAGRANGE_STRAIN_TENSOR","CAUCHY_STRESS_TENSOR","TOTAL_STRESS_TENSOR"]
```
I think this can still be cleaned up. Maybe the entire process can be removed (depending on how the test retrieves things), but at least we only need to output stresses

avdg81 replied:
I have tried to remove the entire GiD output process, but in that case the test fails since retrieving the stresses from the database through the test helper requires a GiD output process to be defined to get the model part:
```
======================================================================
ERROR: test_column_under_gravity (__main__.KratosGeoMechanicsMohrCoulombWithTensionTests.test_column_under_gravity)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Checkouts\KratosProjects\rome\bin\Debug\applications\GeoMechanicsApplication\tests\test_mohr_coulomb_with_tension_cutoff.py", line 73, in test_column_under_gravity
    sig_xx, sig_yy = self.simulate_mohr_coulomb('test_column_under_gravity', 2)
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Checkouts\KratosProjects\rome\bin\Debug\applications\GeoMechanicsApplication\tests\test_mohr_coulomb_with_tension_cutoff.py", line 15, in simulate_mohr_coulomb
    cauchy_stresses = test_helper.get_on_integration_points(simulation, Kratos.CAUCHY_STRESS_TENSOR)
  File "C:\Checkouts\KratosProjects\rome\bin\Debug\applications\GeoMechanicsApplication\tests\test_helper.py", line 308, in get_on_integration_points
    model_part = simulation._list_of_output_processes[0].model_part
                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^
IndexError: list index out of range
```

Therefore, I will limit the requested output to `CAUCHY_STRESS_TENSOR`.

rfaasse replied:
Yep, we need to change that at some point, but for now it's fine

rfaasse reviewed (dismissed):
Looks good! I have a few minor suggestions, but non-blocking, so feel free to either take them into account, or leave them

avdg81 committed:
Processed more review suggestions

- Split a unit test into two tests (to separate the plane strain model from the 3D model).
- Cleaned up the requested output for the new integration test.
- Extended the test description.

rfaasse reviewed (approved):
Good to go!
